rte: use configfile instead of cli args

Tal-or opened the pull request:
As part of bumping RTE -> v0.20.3

We transition RTE to consume its configuration from a configfile (essentially a mounted ConfigMap) instead of commandline arguments.

This change requires adjustments all across the operator code. 
With this PR the operator creates a `ConfigMap` that contains RTE configuration and mounts it into RTE container.
After the bump to v0.19.0+ RTE knows to give precedence to configuration consumed from configfile over a command line args.

Tal-or committed:
vendor: bump rte -> v0.20.3

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
rte:main: use u/s logic for handling configuration
WIP

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

ffromani reviewed (commented):
Nice start.

We have some e2e test which want to consume the RTE config, and/or test RTE command line flags. They likely need to be updated as well:

https://github.com/openshift-kni/numaresources-operator/pull/1075/commits/fa14b5c9cca282ce8b01ffa884f0973c40834a30#diff-49c77cd6ef9fc89514a0030c40a9073e66ef66c94ef0c3ea6a834cecda8ae6ae

Tal-or commented:
We need to fix some security vulnerabilities u/s 
```
Container test exited with code 1, reason Error
---
penshift-kni/numaresources-operator ...
 ✗ [Medium] Path Traversal
   ID: 2fd9a4d3-c857-44e2-8a4b-1f487867022d 
   Path: vendor/github.com/k8stopologyawareschedwg/resource-topology-exporter/pkg/config/cfgfile.go, line 107 
   Info: Unsanitized input from file name flows into os.ReadFile, where it is used as a path. This may result in a Path Traversal vulnerability and allow an attacker to read arbitrary files.
 ✗ [Medium] Path Traversal
   ID: fdd4f27d-afea-4771-b0e8-919f66ff130e 
   Path: vendor/github.com/k8stopologyawareschedwg/resource-topology-exporter/pkg/config/cfgfile.go, line 122 
   Info: Unsanitized input from file name flows into os.ReadFile, where it is used as a path. This may result in a Path Traversal vulnerability and allow an attacker to read arbitrary files.
 ✗ [Medium] Path Traversal
   ID: 6b1a4[96](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift-kni_numaresources-operator/1090/pull-ci-openshift-kni-numaresources-operator-main-security/1861387618982301696#1:build-log.txt%3A96)9-48f1-4ac7-b85e-698fb38437e5 
   Path: vendor/github.com/k8stopologyawareschedwg/resource-topology-exporter/pkg/config/config.go, line 116 
   Info: Unsanitized input from a CLI argument flows into os.ReadDir, where it is used as a path. This may result in a Path Traversal vulnerability and allow an attacker to get a list of arbitrary files.
```

Tal-or committed:
loglevel: `UpdateArgs`

Add a function to convert operator log level
to klog level and set in rte args.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
objectnames: GetDaemonSetConfigName

Add a function that return the DaemonSet config map name.
This configmap used to provide the configuration
for the RTE container runinng under the DaemonSet.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
rteupdate: switch from cli args to configfile

Previously we were setting the configuration for the RTE pod
as command line arguments under the DaemonSet.

We're switching to a config file which means
that the operator sets all configuration in a proper go struct,
stores it in a ConfigMap and mount it into RTE pod.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
rteobjectstate: configmap to store configfile

The operator adds a configmap to hold the configuration for
RTE pod.

The operator makes sure to keep the configmap in the desired state.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
controller: adjust reconciler to use the new logic

Change function names and signature in the main reconciliation logic
to handle RTE via configfile.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
rte: mount configmap

Mount the configmap that holds the config file
into RTE DaemonSet.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
integration-tests: adjust to configfile

instead of checking the DaemonSet container's args
we are checking the arguments inside the injected ConfigMap.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
controller: do not set flags

We should set everything using configfile later
on in the flow.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or commented:
@ffromani 
Security lane is optionally. we don't need to skip

ffromani commented on go.mod:
```diff
@@ -13,7 +13,7 @@ require (
 	github.com/k8stopologyawareschedwg/noderesourcetopology-api v0.1.2
 	github.com/k8stopologyawareschedwg/podfingerprint v0.2.2
-	github.com/k8stopologyawareschedwg/resource-topology-exporter v0.16.1
+	github.com/k8stopologyawareschedwg/resource-topology-exporter v0.19.0
```
please let's consume v0.20.3. Changes between v0.19.0 and v0.20.3 should not create any major disruption: https://github.com/k8stopologyawareschedwg/resource-topology-exporter/compare/v0.19.0...v0.20.3

ffromani reviewed (commented):
I'm not happy with this PR. It's too big of a jump. We need to consume v0.20.3 but keep using the command line args as vehicle to configure RTE. Only later we will switch to configfile.

Tal-or commented:
> I'm not happy with this PR. It's too big of a jump. We need to consume v0.20.3 but keep using the command line args as vehicle to configure RTE. Only later we will switch to configfile.

Too bad we spent so much time working on this one. It would save a lot if would keep the cli args in the first place.
After all this was the original reason for closing #988

ffromani commented:
> > I'm not happy with this PR. It's too big of a jump. We need to consume v0.20.3 but keep using the command line args as vehicle to configure RTE. Only later we will switch to configfile.

The fact we need to narrow down is unfortunate indeed, but we run out of time and the changes are too invasive, and more is likely needed. Reducing the scope is safer and the only option left at this point in time. We will reuse most if not all the code you posted in later updates which we can do incrementally
 
> Too bad we spent so much time working on this one. It would save a lot if would keep the cli args in the first place. After all this was the original reason for closing #988

I guess we could have done a takeover of #988, but the original reason to close it was reprioritization and rebalancing rather than anything else

ffromani commented:
let us reduce further waste: https://github.com/openshift-kni/numaresources-operator/pull/1095

Tal-or commented:
closed due to https://github.com/openshift-kni/numaresources-operator/pull/1090#pullrequestreview-2472867944 
we'll add those changes later on
